src/modelfree/reward_shaping.py

```diff
@@ -0,0 +1,241 @@
+                    self._reset_stat_lists(idx=env_num)
+                    self.results_writer.write_row(epinfo)
+
+                shaped_reward += weighted_reward
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(ResettableAgent):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        - agent: ResettableAgent (most likely) - noise will be added to the actions of
+        this agent in order to build a curriculum of weaker to stronger victims
+        - noise_annealer: Annealer.get_value - presumably the noise should be decreased over time
+        in order to get the adversarial policy to perform well on a normal victim.
+        This function should be tied to a Scheduler to keep it stateless.
+        - noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    def _get_noise_generator(self, noise_type):
+        noise_generators = {
+            'gaussian': lambda x, size: np.random.normal(scale=x, size=size)
+        }
+        return noise_generators[noise_type]
+
+    def get_action(self, observation):
+        noise_param = self.noise_annealer()
+        original_action = self.agent.get_action(observation)
+        action_size = original_action.shape
+
+        noise = self.noise_generator(noise_param, action_size)
+        noisy_action = original_action + noise
+        return noisy_action
+
+    def reset(self):
+        return self.agent.reset()
+
+
+class HumanoidEnvWrapper(VecEnvWrapper):
+    def __init__(self, env):
+        super().__init__(env)
+        self.env = env
+
+    def reset(self):
+        return self.env.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
+
+        return obs, rew, done, infos
+
+
+# TODO: should this inherit from something?
+class Scheduler(object):
+    """Keep track of frac_remaining and return time-dependent values"""
+    schedule_num = 0
+
+    def __init__(self, lr_func=None, rew_shape_func=None, noise_func=None):
+        # print('made a scheduler')
+        # print(Scheduler.schedule_num)
+        self._lr_func = lr_func
+        self._rew_shape_func = rew_shape_func
+        self._noise_func = noise_func
+
+        self.frac_remaining = 1  # frac_remaining goes from 1 to 0
+        self.same_thing = False
```

Comment:
debugging statement. suggest adding `# DEBUG` or `# TODO: remove` or similar so you remember to kill these.